getting-started/compile-contract.md

```diff
@@ -0,0 +1,72 @@
+---
+order: 4
+---
+
+# Download and Compile a Contract
+
+In this section, we will download a sample contract, compile to it to wasm binary executable.
+
+Please first review the [client setup instructions](./setting-env.md), and configure and verify a client, either Go CLI or
+Node.JS console.
+
+## Compiling and Testing a Contract
+
+Let's download the repo which we collect [`cosmwasm-examples`](https://github.com/CosmWasm/cosmwasm-examples) and try out an existing simple escrow contract that can hold some native tokens and gives the power to an arbiter to release them to a pre-defined beneficiary. First clone the repo and try to build the wasm bundle:
+
+```bash
+# get the code
+git clone https://github.com/CosmWasm/cosmwasm-examples
+cd cosmwasm-examples/escrow
+git checkout escrow-0.6.0
+
+# compile the wasm contract with stable toolchain
+rustup default stable
+cargo wasm
+```
+
+After this compiles, it should produce a file in `target/wasm32-unknown-unknown/release/cw_escrow.wasm`. A quick `ls -l` should show around 2MB. This is a release build, but not stripped of all unneeded code. To produce a much smaller
+version, you can run this which tells the compiler to strip all unused code out:
+
+```bash
+RUSTFLAGS='-C link-arg=-s' cargo wasm
+```
+
+This produces a file about 174kB. We use this and another optimizer to produce the final product uploaded to the blockchain.
+You don't need to worry about running this yourself (unless you are curious), but you should have an idea of the final
+size of your contract this way.
+
+For ease of use, you can use [cosmwasm/rust-optimizer](https://github.com/CosmWasm/rust-optimizer) that produces reproducible builds of cosmwasm smart contracts. It also does heavy optimization on the build size, using binary stripping and wasm-opt.
+
+```bash
+docker run --rm -v "$(pwd)":/code \
```

Comment:
Should I provide any more info about this docker command that shows the difference between `RUSTFLAGS='-C link-arg=-s' cargo wasm` @ethanfrey?

Reply:
Just mention it uses `wasm-opt` to strip symbols and debug info for smaller size, but more importantly is a controlled environment that will produce *deterministic builds*.

For local testing the `RUSTFLAGS='-C link-arg=-s' cargo wasm` is fine (and deploying to local dev-nets), but if you want to deploy the contract to a public net, use the docker container, so people can verify your build (prove the source behind it)

Reply:
I would avoid documenting those flags anywhere. We might need to change them and when we do, we want to do that in one place only (rust-optimizer)